transports/webrtc: Implement deterministic certificates generation

thomaseizinger opened the pull request:
# Description

<!-- Please write a summary of your changes and why you made them.-->

This PR implements deterministic certificate generation by means of `ring` and the `simple_x509` library.

Closes https://github.com/libp2p/rust-libp2p/issues/3049.

## Links to any relevant issues

<!-- Reference any related issues.-->


## Open tasks

- https://github.com/rustls/rcgen/pull/166
- https://github.com/webrtc-rs/webrtc/pull/335

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Implement deterministic certificate generation

melekes commented on transports/webrtc/src/tokio/transport.rs:
````diff
@@ -61,20 +61,15 @@ impl Transport {
     /// let transport = Transport::new(id_keys, certificate);
     /// ```
-    pub fn new(id_keys: identity::Keypair, certificate: RTCCertificate) -> Self {
+    pub fn new(id_keys: identity::Keypair, certificate: Certificate) -> Self {
````
Don't think we need to expose Certificate at all if we're able to derive it from `id_keys`

thomaseizinger replied:
We could do that but then it would always be deterministic and would not allow users to put a random one in if they want.

thomaseizinger replied:
Actually, I think I am getting around to this view. I don't think it is necessary for users to put a random certificate in. The more I think about it, the more I'd consider this a footgun actually. Making this an implementation detail gives us more control over the certificate generation and also makes sure we use a proper HKDF to seed the RNG.

thomaseizinger commented:
I can't find any mention of `EDDSA` on https://w3c.github.io/webcrypto/#algorithm-overview so I don't think browsers will support that.

thomaseizinger commented:
I've found https://docs.rs/x509-certificate/0.15.0/x509_certificate/index.html but it doesn't allow to configure the `not_before` parameter which means the certificate will not be deterministic.

thomaseizinger commented:
My investigation so far is that:

- Browsers don't seem to support `EDDSA`.
- `EDDSA` makes deterministic certificates very easy because it doesn't require any additional randomness for the signature.
-  `rcgen` can be used if we could use `EDDSA`.
- `rcgen` can't be used for `ECDSA` because it does not allow you to configure the randomness required for signing. It just instantiates `SystemRandom`.

thomaseizinger commented:
Just found https://github.com/zartarn15/simple_x509 which looks promising from an API PoV.

thomaseizinger committed:
Generate ECDSA certificates instead of EDDSA ones

thomaseizinger commented:
I did it!

We have deterministic ECDSA certificates :partying_face:

melekes commented on transports/webrtc/src/tokio/certificate.rs:
```diff
@@ -0,0 +1,162 @@
+use crate::tokio::fingerprint::Fingerprint;
```
copyright is missing

melekes commented on transports/webrtc/src/tokio/certificate.rs:
```diff
@@ -0,0 +1,162 @@
+                contents: document.as_ref().to_owned(),
+            }),
+            SystemTime::UNIX_EPOCH
+                .checked_add(Duration::from_secs(UNIX_3000 as u64))
```
can't we just use `UNIX_3000`? otherwise 1970 + 3000 years = 4970, correct?

thomaseizinger replied:
Good catch!

thomaseizinger replied:
Actually, the code is correct as is.

`UNIX_3000` represents the year 3000 as a unix timestamp which by definition is the number of seconds since 1970. Thus, if we add `UNIX_3000` to `SystemTime::UNIX_EPOCH`, we get a `SystemTime` for the year 3000.

melekes commented on transports/webrtc/src/tokio/certificate.rs:
```diff
@@ -0,0 +1,162 @@
+    use rand_chacha::ChaCha20Rng;
+
+    #[test]
+    fn certificate_generation_is_deterministic() {
```
could you add a test where two certificates with different seeds are not equal to each other?

melekes reviewed (commented):
![](https://i.giphy.com/media/3ohzdIuqJoo8QdKlnW/giphy.webp)

thomaseizinger commented:
Damn, false alarm! For some reason there is still some hidden randomness somewhere, I just messed up the test.

melekes commented:
can someone with more crypto knowledge also review this? cc @tomaka @mxinden

thomaseizinger committed:
Encode certificate, not private key

thomaseizinger committed:
Make the code more modular

thomaseizinger commented:
This is won't work until https://github.com/briansmith/ring/pull/1279/commits/fe9e4d08c1727bea42d31e0d31ee58c4e8e4dc4e is released.

thomaseizinger commented:
I started by sending a PR to `rcgen`: https://github.com/est31/rcgen/pull/98

tomaka commented:
I haven't really reviewed, but we should be **extremely** careful that the deterministic way in which the certificate is built doesn't change between two versions of libp2p or one of its direct or indirect dependencies.

Since the certificate hash is in the multiaddr, and that you're hardcoding multiaddrs as bootnodes when releasing a version of a peer-to-peer software, the effect of this hash changing would be that someone downloading the software for the first time can't connect anymore.

tomaka commented:
For what it's worth, what I initially had in mind when I proposed deterministically-generated certificates for the spec was to do something like:

```rust
let mut certificate = HARDCODED_CERT_PREFIX.to_vec();
certificate.extend_from_slice(private_key);  // (derived from the libp2p public key in the original idea, but here would be derived from an actual private key)
certificate.extend_from_slice(HARDCODED_CERT_SUFFIX);
```

Where the prefix and suffix are in the spec.

Of course it doesn't actually need to be so strict here, because there's no need for all implementations to agree on how the certificate generation is done, but it would make sure that it's deterministic.

thomaseizinger commented:
> IMO this needs at least a unit test that compares a generated certificate with an expected one, with a huge capital case comment saying "this test shouldn't be removed and do not modify this expected certificate, otherwise code sitting on top of us will break".

I did have that in a previous generation but it wasn't useful in figuring out what function of my generation code is not deterministic. I fully agree with you that need to assert against a hardcoded certificate and I am planning on adding asserting against one as soon as we actually have the deterministic generation working!

> For what it's worth, what I initially had in mind when I proposed deterministically-generated certificates for the spec was to do something like:
> 
> ```rust
> let mut certificate = HARDCODED_CERT_PREFIX.to_vec();
> certificate.extend_from_slice(private_key);  // (derived from the libp2p public key in the original idea, but here would be derived from an actual private key)
> certificate.extend_from_slice(HARDCODED_CERT_SUFFIX);
> ```
> 
> Where the prefix and suffix are in the spec.
> 
> Of course it doesn't actually need to be so strict here, because there's no need for all libp2p implementations to agree on how the certificate generation is done, but it would make sure that it's deterministic.

I am not sure I follow how this is meant to work. Can you explain? Would this certificate be entirely static and the private known to everyone?

The struggle is that ECDSA signature require a random nonce for each signature. If you sign multiple messages with the same private key and not change the nonce, you will leak your private key eventually. If [Schnorr](https://en.wikipedia.org/wiki/Schnorr_signature) wouldn't have patented his signature algorithm, we wouldn't be in this mess but so is life :D

EDDSA also doesn't have this problem but it operates on an entirely different curve and it is not required to be supported by browsers.

Long story short, to deterministically produce an ECDSA signature, you need full control over the randomness source that is used to generate the nonce and that is only possible as of ring 0.17, which is not released yet.

tomaka commented:
> I am not sure I follow how this is meant to work. Can you explain? Would this certificate be entirely static and the private known to everyone?

Sorry got confused. I didn't mean to include the private key in the certificate, but the public key and signature.
It's been ages since I've last looked at what is inside of a certificate, but the basic idea is to hardcode in the source code everything apart from the public key and signature.

> Would this certificate be entirely static and the private known to everyone?

In what I had originally suggested for the WebRTC libp2p spec, yes, the private key would be known to everyone. But that's not what I meant here.

thomaseizinger commented:
> > I am not sure I follow how this is meant to work. Can you explain? Would this certificate be entirely static and the private known to everyone?
> 
> Sorry got confused. I didn't mean to include the private key in the certificate, but the public key and signature.
> It's been ages since I've last looked at what is inside of a certificate, but the basic idea is to hardcode in the source code everything apart from the public key and signature.

There is not much that is mandatory for it to be valid. I think it is only issuer and validity dates.

Is your idea to use hardcoded bytes as a form of template so we don't need to deal with certificate generation libraries? That could work actually!

Unfortunately, if we want to use `ring` for signing, we still have to wait for the next release but we might get away with not depending on `rcgen` then!

tomaka commented:
> That could work actually!

Most importantly, it would guarantee determinism without relying on the implementation details of `rcgen`.
I fear that otherwise a naive `rcgen` contributor could for example do a small change that changes the order of the fields of the certificate, not knowing that it has consequences for us. And the consequences for us would be a silent breakage, which is quite bad.

melekes commented:
> And the consequences for us would be a silent breakage, which is quite bad.

How it's silent if we will have a test that asserts bytes?

tomaka commented:
The test will fail only if you update `rcgen` in the `rust-libp2p` repo.
If `rcgen` publishes a patch version for example, a user can do `cargo update` and get a new version of `rcgen` with a different fields order.

melekes commented:
> The test will fail only if you update `rcgen` in the `rust-libp2p` repo. If `rcgen` publishes a patch version for example, a user can do `cargo update` and get a new version of `rcgen` with a different fields order.

Right, thanks.

thomaseizinger commented:
Stupid me, I already replaced `rcgen` in my latest attempt with `simple_x509` :man_facepalming:

thomaseizinger commented:
> Stupid me, I already replaced `rcgen` in my latest attempt with `simple_x509` man_facepalming

Oh I remember why. It is because `webrtc` also depends on `rcgen` and we need all dependencies to update, even if we are not using the codepath that triggers `rcgen` in webrtc.

thomaseizinger committed:
Temporarily depend on unreleased `ring`

thomaseizinger commented:
The latest patch proves that the deterministic certification generation works and it turns out we don't need to patch `rcgen.` It is enough to patch `webrtc` with the latest `ring` version.

thomaseizinger committed:
Merge branch 'anton/webrtc-transport' into 3049-deterministic-certificates

thomaseizinger committed:
Include copyright header

thomaseizinger committed:
Add test for certificates with different seeds

thomaseizinger committed:
Pin `simple_x509` version

This ensures a patch-bump of the library doesn't accidentally break
our deterministic certificates.

thomaseizinger committed:
Add snapshot test

thomaseizinger commented:
> The test will fail only if you update `rcgen` in the `rust-libp2p` repo. If `rcgen` publishes a patch version for example, a user can do `cargo update` and get a new version of `rcgen` with a different fields order.

I've pinned the `simple_x509` version in the Cargo.toml. To my knowledge, that should prevent this issue.

I toyed around with actually hardcoding the bytes but ASN.1 isn't exactly friendly for that due to this TLV structure. ECDSA public keys and signatures can vary in length, so I would have to compute these lengths and insert them in the correct ASN.1 encoding, then DER-encode the entire thing, run the signature algorithm and build the other x509 ASN.1 structure and serialize all of that to DER.

We could use a ASN.1 library like `yasna` for this but that exposes us to a dependency again. Overall, I think `simple_x509` should to the trick for now. It is an implementation detail so we can always change it to something else.

thomaseizinger committed:
Fix docs

thomaseizinger committed:
Fix imports

thomaseizinger committed:
Move private constants below public API

thomaseizinger committed:
Fix panic message

thomaseizinger committed:
Fix line ending encoding

thomaseizinger committed:
Fix clippy warnings

Ironically, this requires us to change the assertion but the code
isn't shipped yet anyway :D

thomaseizinger committed:
Make certificate deterministic based on identity keys

thomaseizinger committed:
Add docs

thomaseizinger commented:
> > That could work actually!
> 
> Most importantly, it would guarantee determinism without relying on the implementation details of `rcgen`. I fear that otherwise a naive `rcgen` contributor could for example do a small change that changes the order of the fields of the certificate, not knowing that it has consequences for us. And the consequences for us would be a silent breakage, which is quite bad.

I thought more about this.

Should we introduce some kind of identifier / version that users can pass in for how they want their certificate to be generated? In case ring ever ships [RFC6979](https://www.rfc-editor.org/rfc/rfc6979), it would be nice to "upgrade" the certificate generation to use that instead of seeding an RNG. With a bit more effort, we could probably get rid of the `rng` altogether and directly use the output of the HKDF as the the ECDSA private key.

That would be a much cleaner implementation but it will yield a different certificate. Without some kind of version / identifier, we won't be able to provide an upgrade path to users and thus are forever locked into the way we generate certificates.

Thoughts?

thomaseizinger commented:
> > > That could work actually!
> > 
> > 
> > Most importantly, it would guarantee determinism without relying on the implementation details of `rcgen`. I fear that otherwise a naive `rcgen` contributor could for example do a small change that changes the order of the fields of the certificate, not knowing that it has consequences for us. And the consequences for us would be a silent breakage, which is quite bad.
>
> I thought more about this.
> 
> Should we introduce some kind of identifier / version that users can pass in for how they want their certificate to be generated? 
> 
> [snip]
> Without some kind of version / identifier, we won't be able to provide an upgrade path to users and thus are forever locked into the way we generate certificates.
> 
> Thoughts?

I think the upgrade path here is to operate old bootnodes for a while until new versions are shipped that point to new bootnodes that use a different certhash in their multiaddr.

As a result, I don't think we need to ship an upgrade mechanism in code here but leave it up to the application.

melekes commented:
> I think the upgrade path here is to operate old bootnodes for a while until new versions are shipped that point to new bootnodes that use a different certhash in their multiaddr.
> As a result, I don't think we need to ship an upgrade mechanism in code here but leave it up to the application.

I agree. We can just add another method like `generate_from_seed` and have new nodes use this method to derive a certificate.

mxinden commented on transports/webrtc/src/tokio/transport.rs:
````diff
@@ -59,17 +59,22 @@ impl Transport {
     /// ```
-    pub fn new(id_keys: identity::Keypair, certificate: Certificate) -> Self {
-        Self {
+    pub fn new(id_keys: identity::Keypair) -> Result<Self, CertificateError> {
+        let certificate = Certificate::new(&id_keys)?;
````
_Related past discussion https://github.com/melekes/rust-libp2p/pull/12#discussion_r1005204753_

I am not yet decided on whether we should only support deterministic certificate generation. Rational:

- The certificate is valid till the year 3000.
- In case the TLS private key is compromised (leaked, insufficiently secure, ...) an attacker can impersonate the local peer after the Noise handshake with a remote.
- Thus the impersonation attack surface is increased from 1 (libp2p private key) to 2 (libp2p private key + TLS private key).

Related section in the specification:

> - _Why exchange fingerprints in an additional authentication handshake on top of
>   an established WebRTC connection? Why not only exchange signatures of ones TLS
>   fingerprints signed with ones libp2p private key on the plain WebRTC
>   connection?_
> 
>   Once _A_ and _B_ established a WebRTC connection, _A_ sends
>   `signature_libp2p_a(fingerprint_a)` to _B_ and vice versa. While this has the
>   benefit of only requring two messages, thus one round trip, it is prone to a
>   key compromise and replay attack. Say that _E_ is able to attain
>   `signature_libp2p_a(fingerprint_a)` and somehow compromise _A_'s TLS private
>   key, _E_ can now impersonate _A_ without knowing _A_'s libp2p private key.
> 
>   If one requires the signatures to contain both fingerprints, e.g.
>   `signature_libp2p_a(fingerprint_a, fingerprint_b)`, the above attack still
>   works, just that _E_ can only impersonate _A_ when talking to _B_.
> 
>   Adding a cryptographic identifier of the unique connection (i.e. session) to
>   the signature (`signature_libp2p_a(fingerprint_a, fingerprint_b,
>   connection_identifier)`) would protect against this attack. To the best of our
>   knowledge the browser does not give us access to such identifier.

https://github.com/libp2p/specs/tree/master/webrtc#faq

This might be overly cautious. One could argue that people should rotate their libp2p private keys regularly and thus also automatically rotate the TLS private keys.

Also related, people interested in running WebRTC on their bootstrap nodes could as well use the [`dnsaddr`](https://github.com/libp2p/specs/blob/7d934fbd7391d2d567d6d2996cffe6e83af41f7d/addressing/README.md#dnsaddr-links) feature. One would only hardcode the bootstrap node libp2p public key in the client. The client would learn the current TLS certificate fingerprint via the `dnsaddr` TXT record.

What do folks think? In **addition** to the deterministic certificate generation, should we also offer a random certificate code path? Which one should be the default?

melekes replied:
I think providing an option of having deterministic certificate is good. In general there would be two options: 1) generate a new certificate and (optionally) save/load it from disk. 2) derive a certificate from libp2p private key. Users will be free to choose whenever option they see fit. We could add more docs explaining the downsides of each way.

> What do folks think? In addition to the deterministic certificate generation, should we also offer a random certificate code path? Which one should be the default?

If that would simplify UX, sure :+1:

mxinden commented on transports/webrtc/Cargo.toml:
```diff
@@ -16,22 +16,29 @@ bytes = "1"
+rustls = "0.19.1" # Must match version in webrtc library.
 serde = { version = "1.0", features = ["derive"] }
+sha2 = "0.10.6"
+simple_x509 = "=0.2.2" # Version MUST be pinned to ensure a patch-update doesn't accidentially break deterministic certs.
```
This crate has [0 stars on Github](https://github.com/zartarn15/simple_x509),  983 total downloads (none this month) and I don't see any popular repository in the authors GitHub.

I am not saying that this makes someone untrustworthy by itself. I do believe we should have some indicator of trust for all dependencies that we use. Do we have any trust anchor for this crate?

melekes replied:
> Do we have any trust anchor for this crate?

The fact that we're pinning the concrete version and library is small enough for us to review it? Alternatively, we can fork it to `rust-libp2p` if that would increase your trust.

mxinden commented on transports/webrtc/src/tokio/certificate.rs:
```diff
@@ -18,31 +18,68 @@
+use rand::{CryptoRng, Rng, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use ring::signature::{EcdsaKeyPair, KeyPair, ECDSA_P256_SHA256_FIXED_SIGNING};
+use ring::test::rand::FixedSliceSequenceRandom;
```
Is this meant for production code?

melekes replied:
https://github.com/briansmith/ring/blob/48980416c0eecb8e340002b67c33b5040556b7db/src/test.rs#L468-L473

mxinden reviewed (commented):
Thanks for the feature @thomaseizinger.

I still have to dive deeper into the cryptographic primitives used here.

In the meantime, a couple of comments.

@marcopolo given your experience on https://github.com/libp2p/go-libp2p/pull/1833, would you mind taking a look as well?

mxinden commented:
Friendly ping @thomaseizinger. What are the next steps here? Would be unfortunate for the work here to never make it anywhere.

thomaseizinger commented:
I think the proper way of resolving this issue is to use RFC6979, [likely by implementing it for ring](https://github.com/briansmith/ring/issues/1557). That way, we can actually get rid of the randomness altogether and make the signatures properly deterministic.

Anything else I'd consider a hack that is prone to break as soon as implementation details change.

Once ring supports RFC6979 and all dependencies are updated to that ring version, the API we can expose is simply one that takes a bytes slice as seed (and not an RNG). It is then up to the user to fill that byte-slice with random data, use a HKDF, provide a fixed byte array or do whatever else may come to mind.

dariusc93 commented:
Any updates or news on this or would it be best to wait on RFC6979?

/CC @thomaseizinger

EDIT: We could also open a new issue over at rust-libp2p about this (if one doesnt exist already) so there is a consistent point of reference.

DougAnderson444 commented:
> 
> I think we're still waiting for a new [briansmith/ring#1551](https://github.com/briansmith/ring/issues/1551) ring version

0.17.0 [has been released](https://github.com/briansmith/ring/issues/1551#issuecomment-1745701922)
